common/archiver/gcloud/visibilityArchiver.go

```diff
@@ -0,0 +1,259 @@
+		logger.Error(archiver.ArchiveTransientErrorMsg, tag.ArchivalArchiveFailReason(errWriteFile), tag.Error(err))
+		return errRetriable
+	}
+
+	scope.IncCounter(metrics.VisibilityArchiveSuccessCount)
+	return nil
+}
+
+// Query is used to retrieve archived visibility records.
+// Check the Get() method of the HistoryArchiver interface in Step 2 for parameters' meaning and requirements.
+// The request includes a string field called query, which describes what kind of visibility records should be returned. For example, it can be some SQL-like syntax query string.
+// Your implementation is responsible for parsing and validating the query, and also returning all visibility records that match the query.
+// Currently the maximum context timeout passed into the method is 3 minutes, so it's ok if this method takes a long time to run.
+func (v *visibilityArchiver) Query(ctx context.Context, URI archiver.URI, request *archiver.QueryVisibilityRequest) (*archiver.QueryVisibilityResponse, error) {
+	if err := v.ValidateURI(URI); err != nil {
+		return nil, &shared.BadRequestError{Message: archiver.ErrInvalidURI.Error()}
+	}
+
+	if err := archiver.ValidateQueryRequest(request); err != nil {
+		return nil, &shared.BadRequestError{Message: archiver.ErrInvalidQueryVisibilityRequest.Error()}
+	}
+
+	parsedQuery, err := v.queryParser.Parse(request.Query)
+	if err != nil {
+		return nil, &shared.BadRequestError{Message: err.Error()}
+	}
+
+	if parsedQuery.emptyResult {
+		return &archiver.QueryVisibilityResponse{}, nil
+	}
+
+	return v.query(ctx, URI, &queryVisibilityRequest{
+		domainID:      request.DomainID,
+		pageSize:      request.PageSize,
+		nextPageToken: request.NextPageToken,
+		parsedQuery:   parsedQuery,
+	})
+}
+
+func (v *visibilityArchiver) query(ctx context.Context, URI archiver.URI, request *queryVisibilityRequest) (*archiver.QueryVisibilityResponse, error) {
+	token := new(queryVisibilityToken)
+	if request.nextPageToken != nil {
+		var err error
+		token, err = deserializeQueryVisibilityToken(request.nextPageToken)
+		if err != nil {
+			return nil, &shared.BadRequestError{Message: archiver.ErrNextPageTokenCorrupted.Error()}
+		}
+	}
+
+	var prefix = constructVisibilityFilenamePrefix(request.domainID, indexKeyCloseTimeout)
+	if request.parsedQuery.closeTime != 0 {
+		prefix = constructTimeBasedSearchKey(request.domainID, indexKeyCloseTimeout, request.parsedQuery.closeTime, *request.parsedQuery.searchPrecision)
+	}
+	if request.parsedQuery.startTime != 0 {
+		prefix = constructTimeBasedSearchKey(request.domainID, indexKeyStartTimeout, request.parsedQuery.startTime, *request.parsedQuery.searchPrecision)
+	}
+
+	filters := []connector.Precondition{
+		newWorkflowIDPrecondition(hash(common.StringDefault(request.parsedQuery.workflowID))),
+		newRunIDPrecondition(hash(common.StringDefault(request.parsedQuery.runID))),
+		newWorkflowTypeNamePrecondition(hash(common.StringDefault(request.parsedQuery.workflowType))),
```

Comment:
The logic is implemented as part of `hash` function. Inbound string is going to be hashed if it is not empty, Otherwise it will return the given string.

Reply:
I see. But `StringDefault` is called before hash and convert nil pointer to empty string. So basically your assumption here is that if user specified an empty string using query like `workflowID = ""`, it will treated as this filter is not specified right? Is that your desired behavior? If that's the case, can we mention this behavior in README, as it's probably not want people anticipated.  

My suggestion is separate the case for `workflowID = ""` and not specifying workflowID so that the behavior is easier to understand.

Reply:
I agree!. The code looks too much simpler in this way! (easy to understand and read).